pkg/querier/queryrange/shard_resolver.go

```diff
@@ -0,0 +1,153 @@
+	sp, _ := spanlogger.NewWithLogger(r.ctx, r.logger, "dynamicShardResolver.Shards")
+	defer sp.Finish()
+	// We try to shard subtrees in the AST independently if possible, although
+	// nested binary expressions can make this difficult. In this case,
+	// we query the index stats for all matcher groups then sum the results.
+	grps := syntax.MatcherGroups(e)
+
+	// If there are zero matchers groups, we'll inject one to query everything
+	if len(grps) == 0 {
+		grps = append(grps, syntax.MatcherRange{})
+	}
+
+	results := make([]*stats.Stats, 0, len(grps))
+
+	start := time.Now()
+	if err := concurrency.ForEachJob(r.ctx, len(grps), r.maxParallelism, func(ctx context.Context, i int) error {
+		matchers := syntax.MatchersString(grps[i].Matchers)
+		lookback := grps[i].Interval
+		if lookback == 0 {
+			lookback = r.defaultLookback
+		}
+		diff := lookback + grps[i].Offset
+		adjustedFrom := r.from.Add(-diff)
+		adjustedThrough := r.through.Add(-diff)
+
+		start := time.Now()
+		resp, err := r.handler.Do(r.ctx, &indexgatewaypb.IndexStatsRequest{
+			From:     adjustedFrom,
+			Through:  adjustedThrough,
+			Matchers: matchers,
+		})
+		if err != nil {
+			return err
+		}
+
+		casted, ok := resp.(*IndexStatsResponse)
+		if !ok {
+			return fmt.Errorf("expected *IndexStatsResponse while querying index, got %T", resp)
+		}
+
+		results = append(results, casted.Response)
+		level.Debug(sp).Log(
+			"msg", "queried index",
+			"type", "single",
+			"matchers", matchers,
+			"bytes", casted.Response.Bytes,
+			"chunks", casted.Response.Chunks,
+			"streams", casted.Response.Streams,
+			"entries", casted.Response.Entries,
+			"duration", time.Since(start),
+		)
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	combined := stats.MergeStats(results...)
+	factor := guessShardFactor(combined, r.maxParallelism)
+	level.Debug(sp).Log(
+		"msg", "queried index",
+		"type", "combined",
+		"len", len(results),
+		"bytes", combined.Bytes,
+		"chunks", combined.Chunks,
+		"streams", combined.Streams,
+		"entries", combined.Entries,
+		"duration", time.Since(start),
+		"factor", factor,
+	)
+	return factor, nil
+}
+
+const (
+	// Just some observed values to get us started on better query planning.
+	p90BytesPerSecond = 300 << 20 // 300MB/s/core
+	// At max, schedule a query for 10s of execution before
+	// splitting it into more requests. This is a lot of guesswork.
+	maxSeconds          = 10
+	maxSchedulableBytes = maxSeconds * p90BytesPerSecond
+)
+
+func guessShardFactor(stats stats.Stats, maxParallelism int) int {
+	expectedSeconds := float64(stats.Bytes / p90BytesPerSecond)
+	if expectedSeconds <= float64(maxParallelism) {
+		power := math.Ceil(math.Log2(expectedSeconds)) // round up to nearest power of 2
+		// Ideally, parallelize down to 1s queries
+		return int(math.Pow(2, power))
+	}
+
+	n := stats.Bytes / maxSchedulableBytes
+	power := math.Ceil(math.Log2(float64(n)))
+	return int(math.Pow(2, power))
+}
```

Comment:
and here you're just calculating the sharding factor do get some data on what me might want to configure that value to?

Reply:
Yes, we're trying to guess an optimal shard factor based on the amount of data we see in the index.